File: rxdownloadlibrary/src/main/java/rx/downloadlibrary/DownloadProgressResponseBody.java
package rx.downloadlibrary;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import okio.ForwardingSource;
import okio.Okio;
import okio.Source;

public class DownloadProgressResponseBody extends ResponseBody {

    private String identifier;
    private ResponseBody responseBody;
    private DownloadProgressListener progressListener;
    private BufferedSource bufferedSource;

    public DownloadProgressResponseBody(String url, ResponseBody responseBody, DownloadProgressListener progressListener) {
        this.identifier = url;
        this.responseBody = responseBody;
        this.progressListener = progressListener;
    }

    @Override
    public MediaType contentType() {
        return responseBody.contentType();
    }

    @Override
    public long contentLength() {
        return responseBody.contentLength();
    }

    @Override
    public BufferedSource source() {
        if (bufferedSource == null) {
            bufferedSource = Okio.buffer(source(responseBody.source()));
        }
        return bufferedSource;
    }

    private Source source(Source source) {
        return new ForwardingSource(source) {
            long totalBytesRead = 0L;

            @Override
            public long read(Buffer sink, long byteCount) {
                try {
                    long bytesRead = super.read(sink, byteCount);
                    if (bytesRead != -1) {
                        totalBytesRead += bytesRead;
                    }

                    if (progressListener != null) {
                        progressListener.update(identifier, totalBytesRead, responseBody.contentLength(), bytesRead == -1);
                    }
                    return bytesRead;
                } catch (IOException e) {
                    // ignored
                }
                return -1;
            }
        };

    }
}

File: rxdownloadlibrary/src/main/java/rx/downloadlibrary/DownloadProgressResponseBody.kt
package rx.downloadlibrary

import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.*
import java.io.IOException

class DownloadProgressResponseBody(private val url: String, private val responseBody: ResponseBody, private val progressListener: DownloadProgressListener?) : ResponseBody() {

    private var bufferedSource: BufferedSource? = null

    override fun contentType(): MediaType? {
        return responseBody.contentType()
    }

    override fun contentLength(): Long {
        return responseBody.contentLength()
    }

    override fun source(): BufferedSource? {
        if (bufferedSource == null) {
            bufferedSource = Okio.buffer(source(responseBody.source()))
        }
        return bufferedSource
    }

    private fun source(source: Source): Source {
        return object : ForwardingSource(source) {
            var totalBytesRead = 0L

            override fun read(sink: Buffer, byteCount: Long): Long {
                try {
                    val bytesRead = super.read(sink, byteCount)
                    if (bytesRead > -1) {
                        totalBytesRead += bytesRead
                    }
                    //                    bytesRead == -1
                    progressListener?.update(url, totalBytesRead, responseBody.let { contentLength() })
                    return bytesRead
                } catch (e: IOException) {
                    // ignored
                }
                return -1
            }
        }

    }
}
